PGML::Format Refactoring

somiaj opened the pull request:
Refactor the PGML::Format package, which includes changing how it is used, and cleaning up HTML output (including fixing some invalid HTML issues).

+ Refactors how the methods to format the different PGML::Format item methods are called. Instead of using
an odd syntax for switch/cases, `/foo/ && do { ....; last }`, just convert the item `type` into a method name and test if that method exists. Also all item methods are sent the same input, a `state` hash which includes the data they need to determine what output to produce.

+ Fix some issues with invalid HTML, this means that the margin created by `par` blocks are now applied
to the previous list, list item, or indent div, instead of creating an empty div. Also fixed an issue where an `hr` tag was inside a `span` tag.

+ Remove newlines and extra spaces from HTML output. These are extra characters that don't need to be sent to the client.

+ Use `main::tag` to create the HTML tags used in the HTML output.

Based on the code from @dpvc in #1355 and in the review of this PR.

somiaj committed:
Apply Par margin to previous box in some cases in PGML.

Apply the spacing between paragraphs added by the Par block in
in HTML to the previous block instead of inserting an empty div
with a margin. This adds the margin from the Par block to the
previous Indent div, list, or list item block.

Based on the code from @dpvc in #1355.

drgrice1 commented on macros/core/PGML.pl:
```diff
@@ -1266,19 +1266,22 @@ sub format {
-	foreach my $item (@{ $block->{stack} }) {
+	for my $i (0 .. $n - 1) {
+		my $item     = $stack->[$i];
+		my $next_par = $i + 1 < $n && $stack->[ $i + 1 ]{type} eq 'par' ? $stack->[ $i + 1 ] : undef;
```
Change this to
```suggestion
	for my $i (0 .. $#$stack) {
		my $item     = $stack->[$i];
		my $next_par = $i < $#$stack && $stack->[ $i + 1 ]{type} eq 'par' ? $stack->[ $i + 1 ] : undef;
```
and delete `my $n = scalar(@$stack);` above (line 1271).  Perl is optimized for this and stores this number as a property of the array.  So there is no reason to save the array length to yet another location.  Also, perl offers the `$#` syntax for getting the index of the last entry in an array, so use that rather than adding or subtracting one from the array length.

drgrice1 reviewed (dismissed):
This looks good.

I think that the essential thing is that the `div`s that were added between list items be fixed, since that is invalid HTML.  Adding the style to other block items (such as the lists themself) instead of injecting an empty `div` after it is not particularly important since that is at least valid HTML.  I am not entirely certain what benefit that provides either.

somiaj commented:
I am unsure on the benefit as well, but I do feel the empty divs for spacing is not ideal. I was also wondering if there were any accessibility issues with not using `<p>` tags or grouping text. But to me the PGML output is doesn't feel correct.

```html
Paragraph one.
<div style="margin-top:1em"></div>
Paragraph two.
<div style="margin-top:1em"></div>
Paragraph three.
```

And this just feels better to me. But I couldn't figure out a way to do this, and assume that if the PGML parser was setup for this, `<p>` tags would have just been used in the first place.

```html
<div style="margin-bottom:1em">
Paragraph one.
</div>
<div style="margin-bottom:1em">
Paragraph two.
</div>
<div style="margin:0">
Paragraph three.
</div>
```

somiaj commented:
@dpvc Thanks. I am looking at implementing what you stated, but unsure how to modify the `Quote` methods. That method is sent the previous string, not the full block. It appears it is doing a test to determine if it needs to open or close the quote.

dpvc commented:
Sorry, I missed the string being sent to the `Quote` method.  You could add that to the `$self->$method(...)` call arguments.  Alternatively, you could create a `$state` variable that holds any data needed to be passed to the item implementations.  Something like

``` perl
my $state = {
  block => $block,
  strings => [],
  i => 0,
};
while ($state->{i} < $#$stack) {
  my $item = $self->{item} = $stack->[$state->{i}++];
  $self->{nl} = (!defined($state->{strings}[-1]) || $state->{strings}[-1] =~ m/\n$/ ? "" : "\n");
  my $method = ucfirst($item->{type});
  if (Value::can($self, $method)) {
    my $string = $self->$method($item, $state);
    push(@{$state->{strings}}, $string) unless !defined $string || $string eq '';
  } else {
    PGML::Warning "Warning: unknown block type '$item->{type}' in " . ref($self) . "::format\n";
  }
}
```

Then `Quote()` can look in the `$state->{strings}` array, and `Bullet()` can use `$state->{block}` and increment `$state->{i}` if it needs to.  Note that here `$state->{i}` will be pointing to the *next* item in the stack, as it is post-incremented when getting `$item`, so `Bullet()` can use `$state->{i}` rather than `$state->{i} + 1`.

Of course, this requires changes to the implementation functions that previously used `$block` and `$i` as arguments, but there were only a couple of those.

somiaj committed:
Refactor the PGML::format string loop.

As suggested by @dvpc, refactor the PGML::format::string loop to
call the appropriate method if it exists. Also all methods now have
access to the block and current state of the loop to use as needed.

somiaj commented:
Implemented the suggested changes to the main `PGML::format::string` loop as suggested by @dvpc. One difference is I am treating the `$block` and `$state` as two different variables passed to all methods, this way no update is needed to the existing methods that just use the `$block`.

somiaj commented:
Seems my change missed adding a new line character that makes one of the unit tests fails. Trying to track that down unless someone else sees where a newline wasn't added correctly.

drgrice1 reviewed (commented):
While we are fixing HTML validation issues here, there is one more that I have observed.  On line 1638 (in this pull request at least), there is a `span` tag that contains an `hr` tag.  It is invalid for a `span` to contain an `hr`.  That can be easily fixed by changing that `span` to a `div`.  That will not affect the resulting display, and is then valid HTML.

You could add that to this pull request if you like.

drgrice1 commented on macros/core/PGML.pl:
```diff
@@ -1605,7 +1588,7 @@ sub Heading {
 	my $self = shift;
 	my $item = shift;
-	return $self->nl . '<div style="margin-top:1em"></div>' . "\n";
+	return '<div style="margin-top:1em"></div>' . "\n";
```
Yeah, lets leave that for another pull request.

somiaj committed:
Convert span to div to fix HTML validation issue with hr inside.

dpvc commented on macros/core/PGML.pl:
```diff
@@ -1523,17 +1501,15 @@ sub Escape {
-		. "</div>\n";
+	my $em     = 2.25 * $item->{indent};
+	my $bottom = '0';
+	if (defined $block->{stack}[ $state->{i} ] && $block->{stack}[ $state->{i} ]{type} eq 'par') {
```
It might be useful to do

``` perl
my $next = $block->{stack}[ $state->{i} ];
if (defined $next && $next->type eq 'par') {
```

to reduce the number of dereferences that need to be performed.  Similarly for lines 1544 and 1558.

dpvc commented on macros/core/PGML.pl:
```diff
@@ -1523,17 +1501,15 @@ sub Escape {
+		$bottom = '1em';
+		$state->{i}++;
+	}
+	return $self->nl . "<div style=\"margin: 0 0 $bottom ${em}em;\">\n" . $self->string($item) . $self->nl . "</div>\n";
```
If you really are trying to save individual characters being part of the HTML output, then remove the semi-colon after '${em}em' and the `\n` at the end, here.

dpvc commented on macros/core/PGML.pl:
```diff
@@ -1562,20 +1538,27 @@ sub Align {
+	my $list   = $bullet{ $item->{bullet} };
+	my $margin = '0';
+	if (defined $block->{stack}[ $state->{i} ] && $block->{stack}[ $state->{i} ]{type} eq 'par') {
+		$margin = '0 0 1em 0';
```
Again, if you want to reduce characters, leave of the final ` 0`, which is implied by the second `0`.

somiaj committed:
Remove setting PGML::format::item

This was never used anywhere else (that I could find), so no
need to store the current item in the PGML::format::string loop.

dpvc reviewed (dismissed):
I'm OK with this as is, but regarding passing `$block`, I think it makes more sense to put that into the state and modify the few functions that use that.  I almost put `$item` into that as well, but didn't want you to have to edit *every* function.  But there are only a couple that use `$block`.

somiaj committed:
PGML::format combine item and block into state hash.

All of the format methods now receive a single state hash
instead of the item, block, and state as separate inputs.

Reduce the number of times some variables are dereferenced
from the state hash.

Remove some white space to minimize HTML output. Leaving
newline characters `\n` alone for now.

somiaj committed:
PGML::Format make escape method lower case.

This way only "item" methods are capitalized to distinguish
them from other methods which are not a block item type.

somiaj committed:
PGML::Format::html remove new lines and use main::tag.

Since newlines and spaces are extra characters to send, minimize
the HTML output by removing any new line or unneeded space characters.

In addition switch to using `main::tag` to create the html tags.

dpvc commented:
> Do either of you happen to have a problem file which uses all of PGML syntax just to double check I didn't put a typo somewhere?

Unfortunately, no.  I had some PG-based slides from when I helped teach a summer course in WeBWorK authoring, and I did some sessions on PGML and other things like MathObjects.  They had live examples, as I recall, that went through most of the features.  But the PG problems had a lot of version-specific hacks to slim down the user interface and adjust the CSS styling, and I suspect they would no longer work in today's version of WW.  But I could package them up for you if you want them.

Alex-Jordan commented:
Would it help to test by going to the PGML samples from the editor?

somiaj commented:
@Alex-Jordan that is all I am doing. For somethings I looked at the inspector, but mostly just want to make sure the output looks like it should and no warnings appear. I have created a test problem with most common PGML syntax, but I am afraid I might miss some less common setups. Since I didn't change the parser at all (only output), only have to worry that the output looks correct.

somiaj committed:
Use <br> (preferred for HTML5) instead of <br /> in PGML::Format::html.

dpvc reviewed (approved):
Good work on this.  I approve.

drgrice1 reviewed (changes requested):
There is a problem with the `Table` method that is breaking `niceTables.pl` usage.  The `Table` method calls itself with the cell items.  So a "stack" object needs to be passed to its inner call.  If you change line 1326 to `push(@$row, $self->Table({ item => $cell });` that fixes the issue.  The `Table` method doesn't need anything else from the `$state`, so this should be enough.  @dpvc: Can you check this?

Here is a MWE for testing:
```
DOCUMENT();
loadMacros('PGstandard.pl', 'PGML.pl');

BEGIN_PGML
[#
    [.[`x`].]{ headerrow => 1 } [.[`x^2`].]*
    [.[`1`].] [.[`2`].]*
    [.[`2`].] [.[`4`].]*
    [.[`3`].] [.[`9`].]
#]
END_PGML

ENDDOCUMENT();
```

somiaj committed:
Fix issue with PGML::Format::Table.

The table method calls itself and so it also needs to be updated
to send itself a "stack" hash. Since only the `item` is used from
the hash, send only that.

dpvc commented on macros/core/PGML.pl:
```diff
@@ -1266,43 +1266,22 @@ sub format {
+	my $stack = $block->{stack};
+	my @strings;
+	my $state = { block => $block, strings => \@strings, i => 0 };
+	while ($state->{i} <= $#$stack) {
```
About the `$self->Table()` call, I mean.  The `<=` is also a good change.

drgrice1 reviewed (approved):
I think this is good now.
